File: code/packages/typescript/vendor-api-simulators/tests/cross-api.test.ts
/**
 * Cross-API integration tests.
 *
 * These tests verify that all six vendor API simulators can coexist and
 * that the underlying compute runtime is shared correctly.
 */
import { describe, it, expect } from "vitest";
import { halt } from "@coding-adventures/gpu-core";

import {
  // CUDA
  CUDARuntime,
  CUDAMemcpyKind,
  makeCUDAKernel,
  makeDim3,
  // OpenCL
  CLPlatform,
  CLContext,
  CLMemFlags,
  // Metal
  MTLDevice,
  MTLCommandBufferStatus,
  makeMTLSize,
  // Vulkan
  VkInstance,
  VkBufferUsageFlagBits,
  VkSharingMode,
  VkResult,
  VkPipelineBindPoint,
  // WebGPU
  GPU,
  GPUBufferUsage,
  GPUMapMode,
  // OpenGL
  GLContext,
  GL_COMPUTE_SHADER,
  GL_SHADER_STORAGE_BUFFER,
  GL_STATIC_DRAW,
  GL_MAP_READ_BIT,
  // Base
  BaseVendorSimulator,
} from "../src/index.js";

describe("Cross-API: All six simulators instantiate independently", () => {
  it("CUDA creates successfully", () => {
    const cuda = new CUDARuntime();
    expect(cuda._physicalDevice).toBeDefined();
  });

  it("OpenCL creates successfully", () => {
    const ctx = new CLContext();
    expect(ctx._devices.length).toBeGreaterThan(0);
  });

  it("Metal creates successfully", () => {
    const device = new MTLDevice();
    expect(device.name.length).toBeGreaterThan(0);
  });

  it("Vulkan creates successfully", () => {
    const instance = new VkInstance();
    const devices = instance.vkEnumeratePhysicalDevices();
    expect(devices.length).toBeGreaterThan(0);
  });

  it("WebGPU creates successfully", () => {
    const gpu = new GPU();
    const adapter = gpu.requestAdapter();
    expect(adapter.name.length).toBeGreaterThan(0);
  });

  it("OpenGL creates successfully", () => {
    const gl = new GLContext();
    expect(gl._logicalDevice).toBeDefined();
  });
});

describe("Cross-API: BaseVendorSimulator shared patterns", () => {
  it("all simulators extend BaseVendorSimulator", () => {
    expect(new CUDARuntime()).toBeInstanceOf(BaseVendorSimulator);
    expect(new CLContext()).toBeInstanceOf(BaseVendorSimulator);
    expect(new MTLDevice()).toBeInstanceOf(BaseVendorSimulator);
    expect(new VkInstance()).toBeInstanceOf(BaseVendorSimulator);
    expect(new GLContext()).toBeInstanceOf(BaseVendorSimulator);
  });

  it("all simulators have physical and logical devices", () => {
    const sims = [
      new CUDARuntime(),
      new CLContext(),
      new MTLDevice(),
      new VkInstance(),
      new GLContext(),
    ];
    for (const sim of sims) {
      expect(sim._physicalDevice).toBeDefined();
      expect(sim._logicalDevice).toBeDefined();
      expect(sim._computeQueue).toBeDefined();
      expect(sim._memoryManager).toBeDefined();
    }
  });
});

describe("Cross-API: Memory write/read patterns", () => {
  it("CUDA: malloc, memcpy, read back", () => {
    const cuda = new CUDARuntime();
    const ptr = cuda.malloc(8);
    const data = new Uint8Array([1, 2, 3, 4, 5, 6, 7, 8]);
    cuda.memcpy(ptr, data, 8, CUDAMemcpyKind.HostToDevice);
    const result = new Uint8Array(8);
    cuda.memcpy(result, ptr, 8, CUDAMemcpyKind.DeviceToHost);
    expect(result).toEqual(data);
    cuda.free(ptr);
  });

  it("Metal: makeBuffer, writeBytes, contents", () => {
    const device = new MTLDevice();
    const buf = device.makeBuffer(8);
    const data = new Uint8Array([10, 20, 30, 40, 50, 60, 70, 80]);
    buf.writeBytes(data);
    const result = buf.contents();
    expect(result.slice(0, 8)).toEqual(data);
  });

  it("WebGPU: createBuffer, writeBuffer, mapAsync, getMappedRange", () => {
    const gpu = new GPU();
    const adapter = gpu.requestAdapter();
    const device = adapter.requestDevice();
    const buffer = device.createBuffer({
      size: 4,
      usage: GPUBufferUsage.STORAGE | GPUBufferUsage.MAP_READ,
    });
    const data = new Uint8Array([0xaa, 0xbb, 0xcc, 0xdd]);
    device.queue.writeBuffer(buffer, 0, data);
    buffer.mapAsync(GPUMapMode.READ);
    const result = buffer.getMappedRange();
    expect(result).toEqual(data);
    buffer.unmap();
  });

  it("OpenGL: genBuffers, bufferData, mapBufferRange", () => {
    const gl = new GLContext();
    const [buf] = gl.genBuffers(1);
    gl.bindBuffer(GL_SHADER_STORAGE_BUFFER, buf);
    const data = new Uint8Array([5, 10, 15, 20]);
    gl.bufferData(GL_SHADER_STORAGE_BUFFER, 4, data, GL_STATIC_DRAW);
    const result = gl.mapBufferRange(GL_SHADER_STORAGE_BUFFER, 0, 4, GL_MAP_READ_BIT);
    expect(result).toEqual(data);
  });
});

describe("Cross-API: Compute dispatch patterns", () => {
  it("CUDA: kernel launch", () => {
    const cuda = new CUDARuntime();
    const kernel = makeCUDAKernel([halt()], "test");
    cuda.launchKernel(kernel, makeDim3(1, 1, 1), makeDim3(32, 1, 1));
    cuda.deviceSynchronize();
  });

  it("OpenCL: enqueueNDRangeKernel", () => {
    const ctx = new CLContext();
    const queue = ctx.createCommandQueue();
    const prog = ctx.createProgramWithSource("kernel");
    prog.build();
    const kernel = prog.createKernel("compute");
    queue.enqueueNDRangeKernel(kernel, [32]);
    queue.finish();
  });

  it("Metal: command encoder dispatch", () => {
    const device = new MTLDevice();
    const queue = device.makeCommandQueue();
    const cb = queue.makeCommandBuffer();
    const encoder = cb.makeComputeCommandEncoder();
    const lib = device.makeLibrary("src");
    const func = lib.makeFunction("kern");
    const pso = device.makeComputePipelineState(func);
    encoder.setComputePipelineState(pso);
    encoder.dispatchThreadgroups(makeMTLSize(1, 1, 1), makeMTLSize(32, 1, 1));
    encoder.endEncoding();
    cb.commit();
    expect(cb.status).toBe(MTLCommandBufferStatus.completed);
  });

  it("Vulkan: command buffer dispatch", () => {
    const instance = new VkInstance();
    const physicals = instance.vkEnumeratePhysicalDevices();
    const device = instance.vkCreateDevice(physicals[0]);
    const queue = device.vkGetDeviceQueue(0, 0);
    const pool = device.vkCreateCommandPool({ queueFamilyIndex: 0 });
    const [cb] = pool.vkAllocateCommandBuffers(1);

    const shader = device.vkCreateShaderModule({ code: null });
    const dsLayout = device.vkCreateDescriptorSetLayout({ bindings: [] });
    const plLayout = device.vkCreatePipelineLayout({ setLayouts: [dsLayout], pushConstantSize: 0 });
    const [pipeline] = device.vkCreateComputePipelines([{
      shaderStage: { stage: "compute", module: shader, entryPoint: "main" },
      layout: plLayout,
    }]);
    const fence = device.vkCreateFence();

    cb.vkBeginCommandBuffer();
    cb.vkCmdBindPipeline(VkPipelineBindPoint.COMPUTE, pipeline);
    cb.vkCmdDispatch(4, 1, 1);
    cb.vkEndCommandBuffer();

    const result = queue.vkQueueSubmit(
      [{ commandBuffers: [cb], waitSemaphores: [], signalSemaphores: [] }],
      fence,
    );
    expect(result).toBe(VkResult.SUCCESS);
  });

  it("WebGPU: compute pass dispatch", () => {
    const gpu = new GPU();
    const adapter = gpu.requestAdapter();
    const device = adapter.requestDevice();
    const shader = device.createShaderModule({ code: [halt()] });
    const pipeline = device.createComputePipeline({
      layout: "auto",
      compute: { module: shader },
    });
    const encoder = device.createCommandEncoder();
    const pass = encoder.beginComputePass();
    pass.setPipeline(pipeline);
    pass.dispatchWorkgroups(4, 1, 1);
    pass.end();
    const cmdBuf = encoder.finish();
    device.queue.submit([cmdBuf]);
  });

  it("OpenGL: dispatchCompute", () => {
    const gl = new GLContext();
    const shader = gl.createShader(GL_COMPUTE_SHADER);
    gl.shaderSource(shader, "compute");
    gl.compileShader(shader);
    const prog = gl.createProgram();
    gl.attachShader(prog, shader);
    gl.linkProgram(prog);
    gl.useProgram(prog);
    gl.dispatchCompute(4, 1, 1);
    gl.finish();
  });
});

File: code/packages/go/vendor-api-simulators/cross_api_test.go
package vendorapisimulators

import (
	"testing"
)

// =========================================================================
// Cross-API tests -- verify that all six simulators can coexist and that
// shared Layer 5 primitives work correctly across different API wrappers.
// =========================================================================

func TestAllSimulatorsCanBeCreated(t *testing.T) {
	// Every simulator should successfully initialize.
	_, err := NewCUDARuntime()
	if err != nil {
		t.Fatalf("CUDA failed: %v", err)
	}
	_, err = NewCLContext(nil)
	if err != nil {
		t.Fatalf("OpenCL failed: %v", err)
	}
	_, err = NewMTLDevice()
	if err != nil {
		t.Fatalf("Metal failed: %v", err)
	}
	_, err = NewVkInstance()
	if err != nil {
		t.Fatalf("Vulkan failed: %v", err)
	}
	_, err = NewGPUDevice("")
	if err != nil {
		t.Fatalf("WebGPU failed: %v", err)
	}
	_, err = NewGLContext()
	if err != nil {
		t.Fatalf("OpenGL failed: %v", err)
	}
}

func TestAllSimulatorsHavePhysicalDevice(t *testing.T) {
	cuda, _ := NewCUDARuntime()
	if cuda.PhysicalDevice == nil {
		t.Error("CUDA: no physical device")
	}

	cl, _ := NewCLContext(nil)
	if cl.PhysicalDevice == nil {
		t.Error("OpenCL: no physical device")
	}

	mtl, _ := NewMTLDevice()
	if mtl.PhysicalDevice == nil {
		t.Error("Metal: no physical device")
	}

	vk, _ := NewVkInstance()
	if vk.PhysicalDevice == nil {
		t.Error("Vulkan: no physical device")
	}

	wgpu, _ := NewGPUDevice("")
	if wgpu.PhysicalDevice == nil {
		t.Error("WebGPU: no physical device")
	}

	gl, _ := NewGLContext()
	if gl.PhysicalDevice == nil {
		t.Error("OpenGL: no physical device")
	}
}

func TestAllSimulatorsHaveMemoryManager(t *testing.T) {
	cuda, _ := NewCUDARuntime()
	if cuda.MemoryManager == nil {
		t.Error("CUDA: no memory manager")
	}
	cl, _ := NewCLContext(nil)
	if cl.MemoryManager == nil {
		t.Error("OpenCL: no memory manager")
	}
	mtl, _ := NewMTLDevice()
	if mtl.MemoryManager == nil {
		t.Error("Metal: no memory manager")
	}
	vk, _ := NewVkInstance()
	if vk.MemoryManager == nil {
		t.Error("Vulkan: no memory manager")
	}
	wgpu, _ := NewGPUDevice("")
	if wgpu.MemoryManager == nil {
		t.Error("WebGPU: no memory manager")
	}
	gl, _ := NewGLContext()
	if gl.MemoryManager == nil {
		t.Error("OpenGL: no memory manager")
	}
}

func TestAllSimulatorsHaveComputeQueue(t *testing.T) {
	cuda, _ := NewCUDARuntime()
	if cuda.ComputeQueue == nil {
		t.Error("CUDA: no compute queue")
	}
	cl, _ := NewCLContext(nil)
	if cl.ComputeQueue == nil {
		t.Error("OpenCL: no compute queue")
	}
	mtl, _ := NewMTLDevice()
	if mtl.ComputeQueue == nil {
		t.Error("Metal: no compute queue")
	}
	vk, _ := NewVkInstance()
	if vk.ComputeQueue == nil {
		t.Error("Vulkan: no compute queue")
	}
	wgpu, _ := NewGPUDevice("")
	if wgpu.ComputeQueue == nil {
		t.Error("WebGPU: no compute queue")
	}
	gl, _ := NewGLContext()
	if gl.ComputeQueue == nil {
		t.Error("OpenGL: no compute queue")
	}
}

func TestVendorHintConsistency(t *testing.T) {
	// CUDA should always select an nvidia device.
	cuda, _ := NewCUDARuntime()
	if cuda.PhysicalDevice.Vendor() != "nvidia" {
		t.Errorf("CUDA vendor: expected nvidia, got %s", cuda.PhysicalDevice.Vendor())
	}

	// Metal should always select an apple device.
	mtl, _ := NewMTLDevice()
	if mtl.PhysicalDevice.Vendor() != "apple" {
		t.Errorf("Metal vendor: expected apple, got %s", mtl.PhysicalDevice.Vendor())
	}
}

func TestMultipleSimulatorsIndependent(t *testing.T) {
	// Creating multiple simulators should not interfere with each other.
	cuda1, _ := NewCUDARuntime()
	cuda2, _ := NewCUDARuntime()

	// Allocate on cuda1 -- should not affect cuda2.
	ptr, _ := cuda1.Malloc(64)
	cuda1.Free(ptr)

	// cuda2 should still work fine.
	ptr2, err := cuda2.Malloc(128)
	if err != nil {
		t.Fatalf("cuda2 Malloc failed after cuda1 free: %v", err)
	}
	if ptr2.Size != 128 {
		t.Errorf("expected 128, got %d", ptr2.Size)
	}
}

func TestCrossAPIBufferSizes(t *testing.T) {
	// Allocate a buffer through each API and verify size consistency.
	cuda, _ := NewCUDARuntime()
	cudaBuf, _ := cuda.Malloc(256)
	if cudaBuf.Size != 256 {
		t.Errorf("CUDA buffer size: expected 256, got %d", cudaBuf.Size)
	}

	cl, _ := NewCLContext(nil)
	clBuf, _ := cl.CreateBuffer(CLMemReadWrite, 256, nil)
	if clBuf.Size() != 256 {
		t.Errorf("OpenCL buffer size: expected 256, got %d", clBuf.Size())
	}

	mtl, _ := NewMTLDevice()
	mtlBuf, _ := mtl.MakeBuffer(256, MTLResourceStorageModeShared)
	if mtlBuf.Length() != 256 {
		t.Errorf("Metal buffer length: expected 256, got %d", mtlBuf.Length())
	}

	wgpu, _ := NewGPUDevice("")
	wgpuBuf, _ := wgpu.CreateBuffer(GPUBufferDescriptor{Size: 256, Usage: GPUBufferUsageStorage})
	if wgpuBuf.Size() != 256 {
		t.Errorf("WebGPU buffer size: expected 256, got %d", wgpuBuf.Size())
	}
}

func TestCrossAPIDispatchSmoke(t *testing.T) {
	// Each API should be able to dispatch a minimal compute workload.

	// CUDA
	cuda, _ := NewCUDARuntime()
	buf, _ := cuda.Malloc(64)
	err := cuda.LaunchKernel(
		CUDAKernel{Name: "test"},
		NewDim3(1, 1, 1), NewDim3(32, 1, 1),
		[]*CUDADevicePtr{buf}, 0, nil,
	)
	if err != nil {
		t.Errorf("CUDA dispatch failed: %v", err)
	}

	// OpenCL
	cl, _ := NewCLContext(nil)
	prog := cl.CreateProgramWithSource("src")
	prog.Build(nil, "")
	kernel, _ := prog.CreateKernel("test")
	clBuf, _ := cl.CreateBuffer(CLMemReadWrite, 64, nil)
	kernel.SetArg(0, clBuf)
	queue := cl.CreateCommandQueue(nil)
	_, err = queue.EnqueueNDRangeKernel(kernel, []int{32}, []int{32}, nil)
	if err != nil {
		t.Errorf("OpenCL dispatch failed: %v", err)
	}

	// Metal
	mtl, _ := NewMTLDevice()
	lib := mtl.MakeLibrary("src")
	fn := lib.MakeFunction("test")
	pso := mtl.MakeComputePipelineState(fn)
	mtlBuf, _ := mtl.MakeBuffer(64, MTLResourceStorageModeShared)
	mq := mtl.MakeCommandQueue()
	cb, _ := mq.MakeCommandBuffer()
	enc := cb.MakeComputeCommandEncoder()
	enc.SetComputePipelineState(pso)
	enc.SetBuffer(mtlBuf, 0, 0)
	enc.DispatchThreadgroups(NewMTLSize(1, 1, 1), NewMTLSize(32, 1, 1))
	enc.EndEncoding()
	cb.Commit()

	// WebGPU
	wgpu, _ := NewGPUDevice("")
	wenc, _ := wgpu.CreateCommandEncoder()
	pass := wenc.BeginComputePass()
	pipeline := wgpu.CreateComputePipeline(GPUComputePipelineDescriptor{Layout: "auto"})
	pass.SetPipeline(pipeline)
	pass.DispatchWorkgroups(1, 1, 1)
	pass.End()
	wcb, _ := wenc.Finish()
	err = wgpu.Queue.Submit([]*GPUCommandBuffer{wcb})
	if err != nil {
		t.Errorf("WebGPU dispatch failed: %v", err)
	}

	// OpenGL
	gl, _ := NewGLContext()
	shader, _ := gl.CreateShader(GL_COMPUTE_SHADER)
	gl.CompileShader(shader)
	glProg := gl.CreateProgram()
	gl.AttachShader(glProg, shader)
	gl.LinkProgram(glProg)
	gl.UseProgram(glProg)
	glBufs := gl.GenBuffers(1)
	gl.BindBuffer(GL_SHADER_STORAGE_BUFFER, glBufs[0])
	gl.BufferData(GL_SHADER_STORAGE_BUFFER, 64, nil, GL_DYNAMIC_DRAW)
	gl.BindBufferBase(GL_SHADER_STORAGE_BUFFER, 0, glBufs[0])
	err = gl.DispatchCompute(1, 1, 1)
	if err != nil {
		t.Errorf("OpenGL dispatch failed: %v", err)
	}
}
